File: 3394-check-if-grid-can-be-cut-into-sections/3394-check-if-grid-can-be-cut-into-sections.cpp
const int N=200000;
unsigned X[N], Y[N];// pair (x, isStart)-> (x<<1)+isSart
class Solution {
public:
    bool cut3(unsigned* X, int M){
        sort(X, X+M);
        int sum=0, touch0=0;
        for(int i=0; i<M-1; i++){// last end not used
            sum+=(X[i]&1)*2-1;
            touch0+=sum==0;
            if (touch0>=2) return 1;
        }
        return 0;
    }
    bool checkValidCuts(int n, vector<vector<int>>& rectangles) {
        const int m=rectangles.size(), M=m*2;
        int i=0;
        for(auto& r: rectangles){
            unsigned xs=r[0], ys=r[1], xe=r[2], ye=r[3];
            X[i]=(xs<<1)+1; Y[i++]=(ys<<1)+1;
            X[i]=(xe<<1)  ; Y[i++]=(ye<<1); 
        }
        return cut3(X, M) || cut3(Y, M);
    }
};

File: 3394-check-if-grid-can-be-cut-into-sections/3394-check-if-grid-can-be-cut-into-sections.py
class Solution:
    def canCut(self, rectangles, dim):
        rectangles.sort(key = lambda rectangle: rectangle[dim])
        n_sections = 0
        last_end = 0
        for rectangle in rectangles:
            if last_end <= rectangle[dim]:
                n_sections += 1
                if n_sections == 3:
                    return True
            end = rectangle[dim+2]
            if last_end < end:
                last_end = end
        return False
    def checkValidCuts(self, n: int, rectangles: List[List[int]]) -> bool:
        for dim in range(2):
            if self.canCut(rectangles, dim):
                return True
        return False